src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,255 @@
+
+lemma conj_transpose_circulant [has_star α] [add_group I] (v : I → α) :
+  (circulant v)ᴴ = circulant (star (λ i, v (-i))) :=
+by ext; simp [circulant]
+
+lemma fin.transpose_circulant (v : fin n → α) :
+  (circulant v)ᵀ =  circulant (λ i, v (-i)) :=
+begin
+  induction n with n ih, {tidy},
+  simp [transpose_circulant]
+end
+
+lemma fin.conj_transpose_circulant [has_star α] (v : fin n → α) :
+  (circulant v)ᴴ = circulant (star (λ i, v (-i))) :=
+begin
+  induction n with n ih, {tidy},
+  simp [conj_transpose_circulant]
+end
+
+lemma map_circulant [has_sub I] (v : I → α) (f : α → β) :
+  (circulant v).map f = circulant (λ i, f (v i)) :=
+by ext; simp [circulant]
+
+lemma circulant_neg [has_neg α] [has_sub I] (v : I → α) :
+  circulant (- v) = - circulant v :=
+by ext; simp [circulant]
+
+lemma circulant_add [has_add α] [has_sub I] (v w : I → α) :
+  circulant (v + w) = circulant v + circulant w :=
+by ext; simp [circulant]
+
+lemma circulant_sub [has_sub α] [has_sub I] (v w : I → α) :
+  circulant (v - w) = circulant v - circulant w :=
+by ext; simp [circulant]
+
+lemma circulant_mul [semiring α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply],
+  congr' 2,
+  abel
+end
+
+lemma fin.circulant_mul [semiring α] (v w : fin n → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  induction n with n ih, {refl},
+  exact circulant_mul v w,
+end
+
+/-- circulant matrices commute in multiplication under certain condations. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] (v w : fin n → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  induction n with n ih, {refl},
+  exact circulant_mul_comm v w,
+end
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] {k : R} {v : I → α} :
+  circulant (k • v) = k • circulant v :=
+by {ext, simp [circulant]}
+
+lemma circulant_zero [has_zero α] [has_sub I]:
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp [circulant]
+
+/-- The identity matrix is a circulant matrix. -/
+lemma one_eq_circulant [has_zero α] [has_one α] [decidable_eq I] [add_group I]:
+  (1 : matrix I I α) = circulant (pi.single 0 1) :=
```

Comment:
I would flip this around, call it `circulant_single` and make it a `@[simp]` lemma.

Reply:
Should I change the RHS to a different form or just flip around? Did you see the other one right below this?

Reply:
Sure. Will do.